File: src/main/java/org/truffleruby/core/hash/HashOperations.java
/*
 * Copyright (c) 2013, 2023 Oracle and/or its affiliates. All rights reserved. This
 * code is released under a tri EPL/GPL/LGPL license. You can use it,
 * redistribute it and/or modify it under the terms of the:
 *
 * Eclipse Public License version 2.0, or
 * GNU General Public License version 2, or
 * GNU Lesser General Public License version 2.1.
 */
package org.truffleruby.core.hash;

import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
import org.truffleruby.RubyContext;
import org.truffleruby.RubyLanguage;
import org.truffleruby.core.hash.library.EmptyHashStore;
import org.truffleruby.core.hash.library.HashStoreLibrary;
import org.truffleruby.core.numeric.BigIntegerOps;
import org.truffleruby.core.numeric.RubyBignum;
import org.truffleruby.language.RubyBaseNode;

import java.util.Map;
import java.util.Optional;
import java.util.function.Function;

public abstract class HashOperations {

    public static RubyHash newEmptyHash(RubyContext context, RubyLanguage language) {
        return new RubyHash(
                context.getCoreLibrary().hashClass,
                language.hashShape,
                context,
                EmptyHashStore.NULL_HASH_STORE,
                0,
                false);
    }

    @TruffleBoundary
    public static <K, V> RubyHash toRubyHash(RubyContext context, RubyLanguage language,
            HashStoreLibrary hashStoreLibrary, Map<K, V> map, Optional<Function<K, Object>> keyMapper,
            Optional<Function<V, Object>> valueMapper, boolean byIdentity) {
        final RubyHash ret = newEmptyHash(context, language);

        for (Map.Entry<K, V> entry : map.entrySet()) {
            final Object key = keyMapper.isPresent() ? keyMapper.get().apply(entry.getKey()) : entry.getKey();
            final Object value = valueMapper.isPresent() ? valueMapper.get().apply(entry.getValue()) : entry.getValue();

            hashStoreLibrary.set(ret.store, ret, key, value, byIdentity);
        }

        return ret;
    }

    // random number, stops hashes for similar values but different classes being the same, static because we want deterministic hashes
    public static final int BOOLEAN_CLASS_SALT = 55927484;
    public static final int INTEGER_CLASS_SALT = 1028093337;
    public static final int DOUBLE_CLASS_SALT = -1611229937;

    public static long hashBoolean(boolean value, RubyContext context, RubyBaseNode node) {
        return context.getHashing(node).hash(BOOLEAN_CLASS_SALT, Boolean.hashCode(value));
    }

    public static long hashLong(long value, RubyContext context, RubyBaseNode node) {
        return context.getHashing(node).hash(INTEGER_CLASS_SALT, value);
    }

    public static long hashDouble(double value, RubyContext context, RubyBaseNode node) {
        return context.getHashing(node).hash(DOUBLE_CLASS_SALT, Double.doubleToRawLongBits(value));
    }

    public static long hashBignum(RubyBignum value, RubyContext context, RubyBaseNode node) {
        return context.getHashing(node).hash(INTEGER_CLASS_SALT, BigIntegerOps.hashCode(value));
    }
}

File: src/main/ruby/truffleruby/core/truffle/hash_operations.rb
# frozen_string_literal: true

# Copyright (c) 2019, 2025 Oracle and/or its affiliates. All rights reserved. This
# code is released under a tri EPL/GPL/LGPL license. You can use it,
# redistribute it and/or modify it under the terms of the:
#
# Eclipse Public License version 2.0, or
# GNU General Public License version 2, or
# GNU Lesser General Public License version 2.1.

module Truffle
  module HashOperations
    def self.hash_merge(current, other)
      new_hash = current.dup
      other.each_pair do |k, v|
        new_hash[k] = v
      end
      new_hash
    end

    # MRI: rb_hash_set_pair (rb_ary_to_h also contains similar logic for Array#to_h)
    def self.assoc_key_value_pair(hash, pair)
      ary = Truffle::Type.rb_check_convert_type pair, Array, :to_ary

      unless ary
        raise TypeError, "wrong element type #{Primitive.class(pair)} (expected array)"
      end

      if ary.size != 2
        raise ArgumentError, "element has wrong array length (expected 2, was #{ary.size})"
      end

      hash[ary[0]] = ary[1]
    end

    # MRI: extracted from rb_ary_to_h, is similar to rb_hash_set_pair
    def self.assoc_key_value_pair_with_position(hash, pair, index)
      ary = Truffle::Type.rb_check_convert_type pair, Array, :to_ary

      unless ary
        raise TypeError, "wrong element type #{Primitive.class(pair)} at #{index} (expected array)"
      end

      if ary.size != 2
        raise ArgumentError, "wrong array length at #{index} (expected 2, was #{ary.size})"
      end

      hash[ary[0]] = ary[1]
    end
  end
end
